### PR TITLE
Add Json integer, remove specialization.

### PR DESCRIPTION
**Pull request by trivialfis**
Extracted from #4732, which in turns is extracted from #4683 to reduce PR size for quicker review.

### CODE REVIEW

**Comment by trivialfis** [comment]
Integer is used to read pointer value

File: src/common/json.cc [hunk 1]
```diff
@@ -527,15 +525,118 @@ Json JsonReader::ParseObject() {
 }
 
 Json JsonReader::ParseNumber() {
-  std::string substr = raw_str_.substr(cursor_.Pos(), kMaxNumLength);
-  size_t pos = 0;
+  // Adopted from sajson with some simplifications and small optimizations.
```
Comment (hcho3) [inline src/common/json.cc]: Have you considered using `dmlc::strtof()` to parse floating-point numbers?
https://github.com/dmlc/dmlc-core/blob/808f485387f9a03f78fa9f1159f387d0d91b7a28/include/dmlc/strtonum.h#L99

Some differences:
* `dmlc::strtof()` contains more checks for overflow and underflow conditions
* `dmlc::strtof()` supports NAN and INF
* `dmlc::strtof()` avoids the use of `exp10` and `exp` by repeating `*10`: 
https://github.com/dmlc/dmlc-core/blob/808f485387f9a03f78fa9f1159f387d0d91b7a28/include/dmlc/strtonum.h#L184-L199

I think your implementation is reasonable. I'll defer to you to choose between your implementation or `dmlc::strtof()`.

File: src/common/json.cc [hunk 2]
```diff
@@ -527,15 +525,118 @@ Json JsonReader::ParseObject() {
...
+  if (exponent) {
+    CHECK(is_float);
+    // If d is zero but the exponent is huge, don't
+    // multiply zero by inf which gives nan.
+    if (f != 0.0) {
+      // Only use exp10 from libc on gcc+linux
+#if !defined(__GNUC__) || defined(_WIN32) || defined(__APPLE__)
+#define exp10(val) std::pow(10, (val))
+#endif  // !defined(__GNUC__) || defined(_WIN32) || defined(__APPLE__)
+      f *= exp10(exponent);
```
Comment (hcho3) [inline src/common/json.cc]: Is `exp10` faster?

File: src/common/json.cc [hunk 3]
```diff
@@ -585,39 +686,23 @@ class GlobalCLocale {
   }
 };
 
-Json Json::Load(StringView str, bool ignore_specialization) {
+Json Json::Load(StringView str) {
   GlobalCLocale guard;
-  LOG(WARNING) << "Json serialization is still experimental."
```
Comment (hcho3) [inline src/common/json.cc]: Why are we removing this warning?

File: tests/cpp/common/test_json.cc [hunk 4]
```diff
@@ -143,9 +143,26 @@ TEST(Json, TestParseObject) {
 }
 
 TEST(Json, ParseNumber) {
-  std::string str = "31.8892";
-  auto json = Json::Load(StringView{str.c_str(), str.size()});
-  ASSERT_NEAR(get<JsonNumber>(json), 31.8892f, kRtEps);
+  {
```
Comment (hcho3) [inline tests/cpp/common/test_json.cc]: Take a look at https://github.com/dmlc/dmlc-core/blob/808f485387f9a03f78fa9f1159f387d0d91b7a28/test/unittest/unittest_param.cc#L18 and see if we should adopt some of the test cases.

File: tests/cpp/common/test_json.cc [hunk 5]
```diff
@@ -333,18 +351,19 @@ TEST(Json, LoadDump) {
   Json::Dump(origin, &fout);
   fout.close();
 
-  buffer = common::LoadSequentialFile(path);
```
Comment (hcho3) [inline tests/cpp/common/test_json.cc]: I notice that `common::LoadSequentialFile()` has a special logic for POSIX target. Is there significant speed-up when you use `posix_fadvise()` followed by POSIX `read()`?

File: include/xgboost/json.h [hunk 6]
```diff
@@ -256,15 +262,16 @@ class JsonBoolean : public Value {
 };
 
 struct StringView {
```
Comment (hcho3) [inline include/xgboost/json.h]: Aside: we may want to contribute this class into dmlc-core (`dmlc::StringView`)

File: src/common/json.cc (hunk 3 above)
Comment (trivialfis) [reply src/common/json.cc]: @hcho3 Will put it into `Learner::Load` and `Save`.  This implementation will also be used to consume `__array_interface__` from numpy and `__cuda_array_interface__` from cudf and numba and friends.

File: include/xgboost/json.h (hunk 6 above)
Comment (trivialfis) [reply include/xgboost/json.h]: @hcho3 There are quite a lot of requirements to make it really useful.  One from them is `{from, to}_chars` from c++17.  I tried to adopt it from `libcxx` from llvm project, but a few thousands lines of code is unavoidable.

File: tests/cpp/common/test_json.cc (hunk 5 above)
Comment (trivialfis) [reply tests/cpp/common/test_json.cc]: @hcho3 I will remove it in next PR as it's not really useful for loading model from memory buffer, and conflicts with rabit check point logic.

But to answer your question, yes, it's a huge speed up.

File: src/common/json.cc (hunk 1 above)
Comment (trivialfis) [reply src/common/json.cc]: @hcho3 The problem is we also support integer parsing.  It's not in json specification but added by most of implementations.  Also we have lots of integers in tree node index.  So I need a function that can determine whether the string is an integer or a float.

Comment (hcho3) [reply src/common/json.cc]: @trivialfis Got it. In that case, let's just keep `JsonReader::ParseNumber()` as it is.

In the future, when we decide to move your JSON serializer into dmlc-core, we can discuss how to combine the logic in `JsonReader::ParseNumber()` and `dmlc::ParseFloat()` for code de-duplication.

File: src/common/json.cc (hunk 2 above)
Comment (trivialfis) [reply src/common/json.cc]: @hcho3 Sorry I never benchmark this particular function knowing that the newly revised `ParseNumber` is no longer a bottle neck.

**Comment by trivialfis** [comment]
>  I previously read your JSON implementation at #3930.

Ah, that was almost a thousand tickets ago. ;-)

File: tests/cpp/common/test_json.cc (hunk 5 above)
Comment (trivialfis) [reply tests/cpp/common/test_json.cc]: @hcho3 Looking into those tests, JSON number is much simpler.  Like not having `f` suffix, no `inf` or `NAN`.

number

Comment (hcho3) [reply tests/cpp/common/test_json.cc]: @trivialfis Interesting, so NAN and INF are not allowed in JSON. That's awesome

**Comment by trivialfis** [comment]
Merging as no additional logic change to original code base.

**Comment by trivialfis** [comment]
@hcho3 Thanks for the review.

**Pull request merged**